File: flyherder_libs/matlab/flyherder_examples/random_moves.m
function random_moves(port)
% random_moves: performs a homing move followed by a positioning move
% to a random location in the arena as a test of the system.
%
% Note, make sure to set the proper x and y separation when using this
% function. 
%
% Usage: 
%
% random_moves('com5')  % replace com5 with the port of your device.
%
%
    numTrials = 25;
    homeSpeed = 90.0;
    moveSpeed = 50.0;

    maxSeparation.x = 200;
    maxSeparation.y = 200;

    orientation.x0 = '-';
    orientation.y0 = '-';
    orientation.x1 = '+';
    orientation.y1 = '+';

    dev = FlyHerderSerial(port);
    dev.open();
    dev.setDrivePowerOn();

    dev.setMaxSeparation(maxSeparation);  
    dev.setOrientation(orientation);   

    for i = 1:numTrials
        fprintf('trial %d \n',i);
        fprintf('  homing ... ');
        dev.setSpeed(homeSpeed);
        dev.moveToHome();
        dev.wait();  
        fprintf('done\n');

        % Pause for a little bit for added dramma
        pause(1.0)

        % Move to the position
        deltaMax = 0.25*min(maxSeparation.x, maxSeparation.y);
        deltaMin = 0.05*min(maxSeparation.x, maxSeparation.y);
        delta = randi([deltaMin,deltaMax],1);
        pos.x0 = randi([10,maxSeparation.x-(delta+1)],1);
        pos.y0 = randi([10,maxSeparation.y-(delta+1)],1);
        pos.x1 = pos.x0 + delta;
        pos.y1 = pos.y0 + delta;

        fprintf( ... 
            '  moving to position (x0,y0,x1,y1) = (%1.1f,%1.1f,%1.1f,%1.1f) ... ', ... 
            pos.x0, ...
            pos.y0, ...
            pos.x1, ...
            pos.y1  ...
            );

        dev.setSpeed(moveSpeed);
        dev.moveToPosition(pos.x0, pos.y0, pos.x1, pos.y1);
        dev.wait();
        fprintf('done\n');
        pos = dev.getPosition();
        fprintf( ... 
            '  current position   (x0,y0,x1,y1) = (%1.1f,%1.1f,%1.1f,%1.1f)\n', ... 
            pos.x0, ...
            pos.y0, ...
            pos.x1, ...
            pos.y1  ...
            );
    end

    % Clean up - 
    dev.setDrivePowerOff();
    dev.close();
    delete(dev);

end

File: flyherder_libs/python/flyherder_serial/examples/random_moves.py
from __future__ import print_function
import sys
import time
import random
from flyherder_serial import FlyHerder

debug = False
dev = FlyHerder(port='/dev/ttyUSB0',timeout=2.0,debug=debug)


homeSpeed = 90.0
moveSpeed = 30.0
maxSep = 200 

print()
print('homeSpeed {0}'.format(homeSpeed))
print('moveSpeed {0}'.format(moveSpeed))
print('maxSep {0}'.format(maxSep))

dev.setMaxSeparation(maxSep,maxSep)
dev.setOrientation('-','-','+','+')
dev.setDrivePowerOn()

for i in range(0,2):

    print()
    print('trial {0}'.format(i))
    sys.stdout.flush()

    print('  homing ... ',end='')
    sys.stdout.flush()
    dev.setSpeed(homeSpeed)
    dev.moveToHome()
    dev.wait()
    print('done')
    print('  isInHomePosition==', dev.isInHomePosition())

    time.sleep(1.0)

    delta = 20 
    pos_x0 = random.randint(delta+1,int(maxSep - (delta+1)))
    pos_y0 = random.randint(delta+1,int(maxSep - (delta+1)))
    pos_x1 = pos_x0 + delta
    pos_y1 = pos_y0 + delta
    pos = (pos_x0, pos_y0, pos_x1, pos_y1)
    print('  moving to position {0} ... '.format(pos), end='')
    sys.stdout.flush()
    dev.enableBoundsCheck()
    dev.setSpeed(moveSpeed)
    dev.moveToPosition(*pos)
    dev.wait()
    print('done')

    posDict = dev.getPosition()
    posTuple = posDict['x0'], posDict['y0'], posDict['x1'], posDict['y1']
    print('  pos {0}'.format(posTuple))
    sys.stdout.flush()
    time.sleep(1.0)

dev.setDrivePowerOff()



    
